### ISSUE TITLE
[Question] Sort buffers by tab number

### ISSUE THREAD

**Issue by stevanmilic** [akinsho/nvim-bufferline.lua#133]
Hey, is it possible to sort buffers by their respective tabs? e.g. if there's three buffers in tab 1, two buffers in tab 2 and one buffer in tab 3, then they should be shown in-tab order (like I listed them).

I know there's a `sort_by` function I could write, but I'm not sure what's the best way to find tab number from the buffer object.

Thanks for this great plugin!

**Comment by akinsho** [comment]
@stevanmilic you can probably do something like this, but you'd probably need to try and experiment a little. The `sort_by` function gives users free rein to think up and implement any ordering they can think of a solution for.

You could try something like using `vim.api.nvim_tabpage_list_wins()` to get a list of windows, since what a tab really represents is *windows*, not buffers really. Then loop through those and use `vim.api.nvim_win_get_buf(win_id)`. From there, you can maybe assign them a count/order. When you have something like 
```lua
{tab1_buf3, tab1_buf7, tab2_buf2, tab2_buf3}
```
You can then assign each a number and use that to determine the sort order.
In terms of what part of the buffer variable passed in to use, you can `print(vim.inspect(buffer))` so you can use it's `id` which is its bufnr to figure out what index it should have in your list

Anyway hope that helps.

**Comment by stevanmilic** [comment]
Awesome, thanks for the info. I'll try to follow your advice, and post the solution here once I have it.

**Comment by stevanmilic** [comment]
I did it, here's the function:
```lua
sort_by = function(buffer_a, buffer_b)
  local tabs = vim.fn.gettabinfo()
  local buffer_a_tabnr = nil
  local buffer_b_tabnr = nil
  for _, tab in ipairs(tabs) do
      local buffers = vim.fn.tabpagebuflist(tab.tabnr)
      for _, buff in ipairs(buffers) do
          if buff == buffer_a.id then
              buffer_a_tabnr = tab.tabnr
          elseif buff == buffer_b.id then
              buffer_b_tabnr = tab.tabnr
          end
          if buffer_a_tabnr ~= nil and buffer_b_tabnr ~= nil then
              return buffer_a_tabnr < buffer_b_tabnr
          end
      end
  end
  return buffer_a_tabnr < buffer_b_tabnr
end
```
Instead of listing the windows, I just iterate over all tabs, and as soon as I find the buffers I need I make the comparison.

**Comment by akinsho** [comment]
Awesome @stevanmilic I might add this to a wiki section I've been thing of creating to host things like this user created snippets for stuff like `sort_by` and `diagnostics_indicator` etc.

**Comment by stevanmilic** [comment]
Yea that is a great idea @akinsho, I was thinking to suggest the same thing. Even having a support for this out-of-the-box: `sort_by = 'tabs'` would be great. I feel like like this is something users would want to have, since it gives a feeling like you are using tabs instead of buffers, but still handling everything on the buffer level, if you see what I mean.  At that point, a optimization could be made to create a mapping – like you suggested – and access the buffer tab number by a table look-up.

I can try to provide a PR for that, if I find the time.

**Comment by akinsho** [comment]
@stevanmilic I've thought about offering more sorters by default, and there's already a module for that, so it would actually be easy to add. One thing though is, I'm beginning to feel this plugin is getting a little heavy trying to do everything out of the box, whereas by offering users the ability to write custom functions they can do this stuff themselves or copy it from a wiki rather than downloading code they may never want to use.

In the grand scheme of things it's a vanishingly small amount of code, but it's all starting to add up

**Comment by stevanmilic** [comment]
I understand what you're saying, but adding another sorter doesn't make the plugin more-heavy-weight IMO. As long as you develop new functionalities for the concepts you already have, you are not adding complexity,  just adding more options for a feature. And the tests are there to help with the maintenance anyway.

**Comment by akinsho** [comment]
Yeah in this case it's not really a big deal. If you'd be up for a PR I can point you in the right direction. I shouldn't be too hard.
* Add your function to `sorter.lua`
* Add an if clause to match the sorter in the same file (there's a function that handles which sorters)
* Add documentation to the README and help files.
* Add tests for the sorter function to validate it behaves as expected.
